src/ApparentHorizons/StrahlkorperGr.cpp

```diff
@@ -6,15 +6,353 @@
+#include "DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp"
 #include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
 #include "DataStructures/Tensor/Tensor.hpp"
+#include "NumericalAlgorithms/LinearAlgebra/FindGeneralizedEigenvalues.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Christoffel.hpp"
 #include "PointwiseFunctions/GeneralRelativity/IndexManipulation.hpp"
 #include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Gsl.hpp"
 #include "Utilities/MakeWithValue.hpp"
 
+// IWYU pragma: no_forward_declare Tensor
+
+// Functions used by StrahlkorperGr::dimensionful_spin_magnitude
+namespace {
+// Find the 2D surface metric by inserting the tangents \f$\partial_\theta\f$
+// and \f$\partial_\phi\f$ into the slots of the 3D spatial metric
+template <typename Fr>
+tnsr::ii<DataVector, 2, Frame::Spherical<Fr>> get_surface_metric(
+    const tnsr::ii<DataVector, 3, Fr>& spatial_metric,
+    const StrahlkorperTags::StrahlkorperTags_detail::Jacobian<Fr>& tangents,
+    const Scalar<DataVector>& sin_theta) noexcept {
+  auto surface_metric =
+      make_with_value<tnsr::ii<DataVector, 2, Frame::Spherical<Fr>>>(
+          get<0, 0>(spatial_metric), 0.0);
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      get<0, 1>(surface_metric) += spatial_metric.get(i, j) *
+                                   tangents.get(i, 0) * tangents.get(j, 1) *
+                                   get(sin_theta);
+    }
+    // Use symmetry to sum over fewer terms for the 0,0 and 1,1 components
+    get<0, 0>(surface_metric) +=
+        spatial_metric.get(i, i) * square(tangents.get(i, 0));
+    get<1, 1>(surface_metric) += spatial_metric.get(i, i) *
+                                 square(tangents.get(i, 1)) *
+                                 square(get(sin_theta));
+    for (size_t j = i + 1; j < 3; ++j) {
+      get<0, 0>(surface_metric) += 2.0 * spatial_metric.get(i, j) *
+                                   tangents.get(i, 0) * tangents.get(j, 0);
+      get<1, 1>(surface_metric) += 2.0 * spatial_metric.get(i, j) *
+                                   tangents.get(i, 1) * tangents.get(j, 1) *
+                                   square(get(sin_theta));
+    }
+  }
+  return surface_metric;
+}
+
+// Compute the trace of Christoffel 2nd kind on the horizon
+template <typename Fr>
+tnsr::I<DataVector, 2, Frame::Spherical<Fr>> get_trace_christoffel_second_kind(
+    const tnsr::ii<DataVector, 2, Frame::Spherical<Fr>>& surface_metric,
+    const tnsr::II<DataVector, 2, Frame::Spherical<Fr>>& inverse_surface_metric,
+    const Scalar<DataVector>& sin_theta, const YlmSpherepack& ylm) noexcept {
+  const Scalar<DataVector> cos_theta{cos(ylm.theta_phi_points()[0])};
+
+  // Because the surface metric components are not representable in terms
+  // of scalar spherical harmonics, you can't naively take first derivatives.
+  // To avoid a potential coordinate singularity, actually differentiate
```

Comment:
I guess now that you mention it, it might not technically be singular coordinates, just bad behavior with naive derivatives. I'll reword when I squash to read "To avoid potential numerical error," instead.